arb_os/evmCallStack.mini

```diff
@@ -322,81 +320,91 @@ public impure func evmCallStack_doCall(
-
-    // prepare the caller's frame
-    //    this means (1) saving the runningAsAccount into the accountStore, and
-    //               (2) saving the resumeInfo so we can resume later
-    let stackContents = saveAvmStack();  // must be done as a top level statement
-    topFrame = topFrame
-        with { accountStore: accountStore_set(topFrame.accountStore, callerAddr, caller) }
-        with { resumeInfo: Some(struct {
-                   codePoint: resumeCodePoint,
-                   stackContents: stackContents,
-                   returnDataOffset: retOffset,
-                   returnDataNbytes: retLength,
-                   callType: kind,
-                   savedGas: availableGas-gas,
-               }) };
-
-    // now create a new callframe for the callee
-    //     it incorporates the balance transfer from caller to callee
-    //          (which is put into the callee frame so it isn't committed unless the call succeeds)
     let callee = accountStore_get(topFrame.accountStore, calleeAddr);
-    let runningAsAddr = calleeAddr;
-    if (kind == 2) {
-        runningAsAddr = callerAddr;  // delegatecall runs as the caller, with callee's code
-    }
-    let updatedAcctStore = topFrame.accountStore;
-    if (kind == 0) {  //TODO: verify that txcall is the only one that is supposed to allow payment-with-call
-        let updatedAcctStore = accountStore_transferEthBalance(
-            updatedAcctStore,
-            callerAddr,
-            calleeAddr,
-            balance
-        )?;
-    }
-    let newTopFrame = struct {
-        runningAs: calleeAddr,
-        runningCodeFrom: calleeAddr,
-        accountStore: updatedAcctStore,
-        runningAsAccount: accountStore_get(updatedAcctStore, calleeAddr),
-        caller: topFrame.runningAs,
-        callKind: kind,
-        calldata: bytearray_extract(topFrame.memory, argsOffset, argsLength),
-        callvalue: balance,
-        returnInfo: None<ReturnInfo>,
-        memory: bytearray_new(0),
-        evmLogs: topFrame.evmLogs,
-        selfDestructQueue: topFrame.selfDestructQueue,
-        resumeInfo: None<ResumeInfo>,
-        parent: Some(topFrame)
-    };
+    if let Some(startCodePoint) = account_getStartCodePoint(callee) {
+        // Create a new callframe for a subcall, and transfer execution to it.
+        let availableGas = gasAccounting_pauseTxCharges();
+        if (gas > availableGas) {
+            gas = availableGas;  // don't give the callee more gas than the caller has
+        }
+
+        if (kind == 1) {
+            errorHandler();  // should never return
+            panic;  // callcode call type not yet implemented
+        }
 
-    // there were no errors in the setup, so write the result to global storage
-    globalCallStack = Some(newTopFrame);
+        let callerAddr = topFrame.runningAs;
+        if (callerAddr == calleeAddr) {
+            return None;  // a contract shouldn't use the external call mechanism to call itself
+        }
```

Comment:
A contract should actually be able to do this

Reply:
Removed this code.